docs/design/guideline_recategorization.md

```diff
@@ -0,0 +1,178 @@
+| Required   | Violations are reported unless there exists an applicable deviation. |
+| Advisory   | Violations are reported unless there exists an applicable deviation. |
+| Disapplied | Violations are not reported.                                         |
+
+The SARIF rewriting will update the category of a guideline in a SARIF result file by updating the necessary tag information of a query.
+
+### Guideline Recategorization Plan specification
+
+The Guideline Recategorization Plan specification will build upon the configuration specification introduced for deviations by adding the additional primary section `guideline-recategorizations` to the `codeql-standards.yml` configuration file.
+The `guideline-recategorizations` section will be a series of compact mappings in YAML with the keys:
+
+- `rule-id` - the rule identifier that is recategorized.
+- `category` - the category assigned to the rule identified by rule-id
+
+Note: We specify the recategorization based on the rule-id instead of the query-id. This can be revised if feedback requires more fine-grained recategorization.
+
+For example:
+
+```yaml
+guideline-recategorizations:
+- rule-id: “M5-0-17”
+  category: “mandatory”
+```
+
+## Implementation
+
+This section discusses the implementation of the [design](#design).
+
+### Specification and deviation
+
+The implementation will rely on the existing rule meta-data and query exclusion mechanisms to apply policies associated with a rule’s category.
+The rule meta-data already includes both the `query-id` and `rule-id` associated with a query and is available during query evaluation.
+The rule meta-data needs to be extended with a category that contains the guideline’s category.
+
+For example:
+
+```ql
+ query =
+    // `Query` instance for the `pointerSubtractionOnDifferentArrays` query
+    PointersPackage::pointerSubtractionOnDifferentArraysQuery() and
+  queryId =
+    // `@id` for the `pointerSubtractionOnDifferentArrays` query
+    "cpp/autosar/pointer-subtraction-on-different-arrays" and
+  ruleId = "M5-0-17" and
+  category = “required”
+```
+
+The category defined by the rule meta-data and the category defined in the `guideline-recategorizations` of the applicable `codeql-standards.yml` configuration file is used to determine the *effective category* of a query.
+The *effective category* is the category whose policy is applied during the evaluation of a query.
+The policy of a category dictates if a result can be deviated from and implements the effect described in the design section.
+The existing exclusion mechanism implemented in the predicate `isExcluded` defined in the `Exclusions.qll` library will be updated to consider the applicable policy of a guideline.
+
+Note: This will change the behavior of deviations which will no longer have an impact on Mandatory guidelines! This, however, will only impact MISRA C rules because there are no MISRA C++ Guidelines with a Mandatory category.
```

Comment:
```suggestion
Note: This changes the behavior of deviations which will no longer have an impact on Mandatory guidelines! However, this will only affect MISRA C rules because there are no MISRA C++ Guidelines with a Mandatory category.
```

Reply:
Applied